### PR TITLE
Support displaying diff stats in PR tab bar

### LINKED ISSUE #25326: Diff stats in PR tab bar
### Feature Description

It would be nice if we could show a diff stat on the right side of the PR tab bar.

### Screenshots

image

### PR DESCRIPTION
**Pull request by hiifong**
Fix #25326 

image

### CODE REVIEW

**Comment by silverwind** [comment]
Align it to far-right? I assume a `margin-left: auto` shall do.

**Comment by hiifong** [comment]
> Align it to far-right? I assume a `margin-left: auto` shall do.

I don't know how to adjust the css style, can you help me?

**Comment by silverwind** [comment]
Yeah I can try to adjust it later.

**Comment by silverwind** [comment]
Styling enhancement done:

image

File: templates/repo/pulls/tab_menu.tmpl [hunk 1]
```diff
@@ -14,4 +14,10 @@
 		{{$.locale.Tr "repo.pulls.tab_files"}}
 		<span class="ui small label">{{if .NumFiles}}{{.NumFiles}}{{else}}-{{end}}</span>
 	</a>
+	<span class="item gt-ml-auto gt-pr-0 gt-font-bold gt-df gt-ac gt-gap-3">
+		<span><span class="text green">+{{if .Diff.TotalAddition}}{{.Diff.TotalAddition}}{{else}}0{{end}}</span> <span class="text red">-{{if .Diff.TotalDeletion}}{{.Diff.TotalDeletion}}{{else}}0{{end}}</span></span>
+		<span class="diff-stats-bar">
+			<div class="diff-stats-add-bar" style="width: 50%;"></div>
```
Comment (silverwind) [inline templates/repo/pulls/tab_menu.tmpl]: This percentage is still pending?

Comment (hiifong) [reply templates/repo/pulls/tab_menu.tmpl:20]: Confirmed

Comment (hiifong) [reply templates/repo/pulls/tab_menu.tmpl:20]: Need to convert to css class？Need your help

Comment (silverwind) [reply templates/repo/pulls/tab_menu.tmpl]: Use `Eval` like done here:

https://github.com/go-gitea/gitea/blob/a95a7bcdf30505150c8c94e1700599fcfad9f98e/templates/repo/diff/stats.tmpl#L4

File: routers/web/repo/pull.go [hunk 2]
```diff
@@ -357,6 +357,72 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+	}
+	diffOptions := &gitdiff.DiffOptions{
+		BeforeCommitID:     startCommitID,
+		AfterCommitID:      endCommitID,
+		SkipTo:             ctx.FormString("skip-to"),
+		MaxLines:           maxLines,
+		MaxLineCharacters:  setting.Git.MaxGitDiffLineCharacters,
+		MaxFiles:           maxFiles,
+		WhitespaceBehavior: gitdiff.GetWhitespaceFlag(ctx.Data["WhitespaceBehavior"].(string)),
+	}
```
Comment (wxiaoguang) [inline routers/web/repo/pull.go]: I guess the code is just copied & pasted, I do not think it looks good for this feature. For example, what are the "file-only", "files", "skip-to" for?

File: templates/repo/pulls/tab_menu.tmpl (hunk 1 above)
Comment (hiifong) [reply templates/repo/pulls/tab_menu.tmpl]: Is this expected?
image

Comment (silverwind) [reply templates/repo/pulls/tab_menu.tmpl]: yes 80% would be expected here.

File: routers/web/repo/pull.go [hunk 3]
```diff
@@ -356,6 +356,64 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+	var (
+		startCommitID string
+		endCommitID   string
+		gitRepo       = ctx.Repo.GitRepo
+	)
+
+	var prInfo *git.CompareInfo
+
+	if pull.HasMerged {
+		prInfo = PrepareMergedViewPullInfo(ctx, issue)
```
Comment (lunny) [inline routers/web/repo/pull.go]: We just need `prInfo.MergeBase` but these two functions are too heavy for that.

File: routers/web/repo/pull.go [hunk 4]
```diff
@@ -356,6 +356,64 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+
+	var (
+		startCommitID string
+		endCommitID   string
+		gitRepo       = ctx.Repo.GitRepo
+	)
+
+	var prInfo *git.CompareInfo
+
+	if pull.HasMerged {
```
Comment (hiifong) [inline routers/web/repo/pull.go]: Do we really need it?

**Review by silverwind — APPROVED** [review]
Can't speak for backend changes, but frontend looks good.

File: routers/web/repo/pull.go [hunk 5]
```diff
@@ -356,12 +356,62 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+		return
+	}
+
+	headCommitID, err := gitRepo.GetRefCommitID(pull.GetGitRefName())
+	if err != nil {
+		ctx.ServerError("GetRefCommitID", err)
+		return
+	}
+
+	startCommitID = mergeBaseCommitID
```
Comment (lunny) [inline routers/web/repo/pull.go]: We don't need this two local variables.

File: routers/web/repo/pull.go [hunk 6]
```diff
@@ -356,12 +356,62 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+		return
+	}
+
+	ctx.Data["Diff"] = diff
+}
+
+func GetMergedBaseCommitID(ctx *context.Context, issue *issues_model.Issue) string {
 	pull := issue.PullRequest
 
 	setMergeTarget(ctx, pull)
```
Comment (lunny) [inline routers/web/repo/pull.go]: We don't need this line.

File: services/gitdiff/gitdiff.go [hunk 7]
```diff
@@ -1229,6 +1229,42 @@ func GetDiff(gitRepo *git.Repository, opts *DiffOptions, files ...string) (*Diff
...
+	repoPath := gitRepo.Path
+
+	diff := &PullDiffStats{}
+
+	separator := "..."
+	if opts.DirectComparison {
+		separator = ".."
+	}
+
+	diffPaths := []string{opts.BeforeCommitID + separator + opts.AfterCommitID}
```
Comment (lunny) [inline services/gitdiff/gitdiff.go:1247]: Could we use `gitRepo.GetDiffShortStat` directly?

Comment (hiifong) [reply services/gitdiff/gitdiff.go:1247]: This looks like it shouldn't be possible

Comment (hiifong) [reply services/gitdiff/gitdiff.go:1247]: I'm not sure

**Comment by hiifong** [comment]
We need to ignore the diff stats of merge pr?

File: routers/web/repo/pull.go [hunk 8]
```diff
@@ -356,12 +356,58 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+// GetPullDiffStats get Pull Requests diff stats
+func GetPullDiffStats(ctx *context.Context) {
+	issue := checkPullInfo(ctx)
 	pull := issue.PullRequest
 
-	setMergeTarget(ctx, pull)
-	ctx.Data["HasMerged"] = true
+	var (
+		endCommitID string
+		gitRepo     = ctx.Repo.GitRepo
```
Comment (lunny) [inline routers/web/repo/pull.go]: Both these two variables could be removed.

File: routers/web/repo/pull.go [hunk 9]
```diff
@@ -356,12 +356,58 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+		MaxLines:           setting.Git.MaxGitDiffLines,
+		MaxLineCharacters:  setting.Git.MaxGitDiffLineCharacters,
+		MaxFiles:           setting.Git.MaxGitDiffFiles,
+		WhitespaceBehavior: gitdiff.GetWhitespaceFlag(ctx.Data["WhitespaceBehavior"].(string)),
+	}
+
+	var methodWithError string
+	var diff *gitdiff.PullDiffStats
+
+	diff, err = gitdiff.GetPullDiffStats(gitRepo, diffOptions)
```
Comment (lunny) [inline routers/web/repo/pull.go]:
```suggestion
	diff, err := gitdiff.GetPullDiffStats(gitRepo, diffOptions)
```

File: routers/web/repo/pull.go [hunk 10]
```diff
@@ -356,12 +356,58 @@ func setMergeTarget(ctx *context.Context, pull *issues_model.PullRequest) {
...
+	methodWithError = "GetPullDiffStats"
+
+	if err != nil {
+		ctx.ServerError(methodWithError, err)
```
Comment (lunny) [inline routers/web/repo/pull.go]:
```suggestion
	if err != nil {
		ctx.ServerError("GetPullDiffStats", err)
```

**Comment by silverwind** [comment]
Make sure to "Resolve" all resolved discussion threads.

**Comment by hiifong** [comment]
https://github.com/go-gitea/gitea/pull/25387#pullrequestreview-1502129719 is not sure

**Review by lunny — APPROVED** [review]

**Pull request merged**